src/query/interpret/eval.hpp

```diff
@@ -479,9 +479,37 @@ class ExpressionEvaluator : public ExpressionVisitor<TypedValue> {
       case TypedValue::Type::Null:
         return TypedValue(ctx_->memory);
       case TypedValue::Type::Vertex:
-        return TypedValue(GetProperty(expression_result_ptr->ValueVertex(), property_lookup.property_), ctx_->memory);
+        if (property_lookup.evaluation_mode_ == PropertyLookup::EvaluationMode::GET_ALL_PROPERTIES) {
+          auto symbol_pos = static_cast<Identifier *>(property_lookup.expression_)->symbol_pos_;
+          if (!ctx_->property_lookups_cache.contains(symbol_pos)) {
+            ctx_->property_lookups_cache[symbol_pos] = GetAllProperties(expression_result_ptr->ValueVertex());
+          }
+
+          auto property_id = ctx_->properties[property_lookup.property_.ix];
+          if (!ctx_->property_lookups_cache.at(symbol_pos).contains(property_id)) {
+            return TypedValue(ctx_->memory);
+          }
+
+          return TypedValue(ctx_->property_lookups_cache.at(symbol_pos).at(property_id), ctx_->memory);
+        } else {
+          return TypedValue(GetProperty(expression_result_ptr->ValueVertex(), property_lookup.property_), ctx_->memory);
+        }
       case TypedValue::Type::Edge:
-        return TypedValue(GetProperty(expression_result_ptr->ValueEdge(), property_lookup.property_), ctx_->memory);
+        if (property_lookup.evaluation_mode_ == PropertyLookup::EvaluationMode::GET_ALL_PROPERTIES) {
+          auto symbol_pos = static_cast<Identifier *>(property_lookup.expression_)->symbol_pos_;
+          if (!ctx_->property_lookups_cache.contains(symbol_pos)) {
+            ctx_->property_lookups_cache[symbol_pos] = GetAllProperties(expression_result_ptr->ValueEdge());
+          }
+
+          auto property_id = ctx_->properties[property_lookup.property_.ix];
+          if (!ctx_->property_lookups_cache.at(symbol_pos).contains(property_id)) {
+            return TypedValue(ctx_->memory);
+          }
+
+          return TypedValue(ctx_->property_lookups_cache.at(symbol_pos).at(property_id), ctx_->memory);
+        } else {
+          return TypedValue(GetProperty(expression_result_ptr->ValueEdge(), property_lookup.property_), ctx_->memory);
+        }
```

Comment:
I would rather reverse the ifs. have the getting of an own property as a first if which can hit, because if we fall to getting all properties, we already have an optimization there.

Reply:
Don't know whether that would work. The first `if` block makes sure that properties are cached, and the second `if` block, which basically checks whether a property is nonexistent, must follow it.